### ISSUE TITLE
Weird RelationalPred rule behavior with ClassRef

### ISSUE THREAD

**Issue by OrangeNote** [catseye/ALPACA#5]
This works as intended, X doesn't change its state because it doesn't belong to class C:
```
class A;
class C;

state Blank " ";
state X "*" is A to Blank when me is C

begin
*
```
Result: `* -> *`

But this doesn't work! State X is A and also B, but not C, so it shouldn't evolve to Blank, but it does.
```
class A is B;
class B;
class C;

state Blank " ";
state X "*" is A to Blank when me is C

begin
*
```
Result: `* ->  `

**Comment by cpressey** [comment]
Confirmed there is some sort of problem here.

It seems it's not actually evolving to Blank, because if you change the definition of Blank to `state Blank "Z"`, it does not show a `Z`, it still shows a blank space.

Also, neither of these examples compile to Javascript: node complains `is_C is not defined`.

I see the same behaviour if I replace `me` with `^v`.

**Comment by cpressey** [comment]
It occurred to me that in both these examples, C has no member states. The predicate `is C` will always be false.

I haven't looked, but it seems very likely that the code assumes somewhere that a class has at least one member, and the breaking of that assumption is causing this bug.

It also occurred to me that it would also be possible to statically prohibit empty classes like this, i.e. make it a compile-time error to specify a class but to have no states in that class.  That would eliminate this situation entirely. But I'm not sure it's entirely justified, from a language-design perspective. I'll think about it.

**Comment by cpressey** [comment]
OK, I retract one of my observations in https://github.com/catseye/ALPACA/issues/5#issuecomment-360215953 - I added some debugging (in a new branch, [here](https://github.com/catseye/ALPACA/commit/25f250d407f7543c8f30ab33530376858292e1e8)) and it is indeed evolving to Blank.  I had forgotten that the first state is always the "empty" state, and even if you make it look like something, it won't necessarily be displayed (because it's "everywhere anyway").

Still, there is a display issue too, because what I see for an "entirely blank" playfield is

    -----
     -----

I definitely do not expect the space on the 2nd line.

**Comment by cpressey** [comment]
I see the problem. `alpaca.analysis.state_defn_is_a` is basically looking at the wrong class, when it checks to see if the state is a member of the class.  (It considers every class that is declared in the ALPACA file, in a loop, and it is checking _those_ classes for membership in other classes.  It seems, if _any_ class descends from C, an `is C` test will pass.  That explains the discrepancy between the two example programs in the bug report.)

I'll work on a fix over the next little while.

**Comment by cpressey** [comment]
@OrangeNote The bug itself is now fixed in [this commit](https://github.com/catseye/ALPACA/commit/e8e70059de97f88ba78a8cc87e436aa5a361ba90).

I've added your example with unexpected behaviour as a test case in [this commit](https://github.com/catseye/ALPACA/commit/fcd5a8bdd579c209475dedd2e221b8c437d5ca21), which passes when the above bug fix is applied.

Both of these changes are in the [develop-1.1 branch](https://github.com/catseye/ALPACA/compare/develop-1.1) and will appear in the release of version 1.1.

Thanks for your report!

(The issues with Javascript compilation, and strange display of an empty playfield, are still outstanding, so I'll leave this issue open for now.)
